### PR TITLE
Add caching for metadata store queries

### LINKED ISSUE #981: Apply caching to metadatastore.js
We should cache the queries that are sent to metadatastore.js to improve the read performance of the addon now that we are storing all of our metadata there.  We need to make sure that we consider the performance of the metadata database due to deduping.

### PR DESCRIPTION
**Pull request by ncloudioj**
Fixes #981 

This change is [Reviewable](https://reviewable.io/reviews/mozilla/activity-stream/1064)

### CODE REVIEW

**Comment by ncloudioj** [comment]
WIF, will fix #981

**Comment by oyiptong** [comment]
Review status: 0 of 7 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[package.json, line 148 \[r1\]](https://reviewable.io:443/reviews/mozilla/activity-stream/1064#-KOpX8x0F6TUWcjqqDlM:-KOpX8x0F6TUWcjqqDlN:b-npxaov) ([raw file](https://github.com/mozilla/activity-stream/blob/8e75fd7690d9625d17f2b217122bce0f7076e998/package.json#L148)):*
> ```JSON
>     {
>       "name": "as-simple-storage.writePeriod",
>       "title": "as-simple-storage wirte period",
> ```

typo: write not wirte

Also, nit: you might want a better description while you're at it, even if it's hidden

---

*[package.json, line 155 \[r1\]](https://reviewable.io:443/reviews/mozilla/activity-stream/1064#-KOpXOXpop8yYrhxlRNa:-KOpXOXpop8yYrhxlRNb:b-ep8p7d) ([raw file](https://github.com/mozilla/activity-stream/blob/8e75fd7690d9625d17f2b217122bce0f7076e998/package.json#L155)):*
> ```JSON
>     {
>       "name": "as-simple-storage.quota",
>       "title": "as-simple-storage quota",
> ```

As above... something more descriptive may be nice?

e.g. "disk quota for activity stream's persistent in-memory storage"

---

*[lib/SimpleStorage.js, line 25 \[r1\]](https://reviewable.io:443/reviews/mozilla/activity-stream/1064#-KOpXqFCbet_1oK_mi2k:-KOpXqFDDLxVrd-Wkrpc:b-ix7xgs) ([raw file](https://github.com/mozilla/activity-stream/blob/8e75fd7690d9625d17f2b217122bce0f7076e998/lib/SimpleStorage.js#L25)):*
> ```JavaScript
> 
> const SIMPLE_STORAGE_WRITE_PERIOD = "as-simple-storage.writePeriod";
> // const WRITE_PERIOD_DEFAULT = 300000; // 5 minutes
> ```

please remove comment

---

*[lib/SimpleStorage.js, line 28 \[r1\]](https://reviewable.io:443/reviews/mozilla/activity-stream/1064#-KOpXsdImu0qtznYCJbm:-KOpXsdJTJWTgmgEbviY:b-ix7xgs) ([raw file](https://github.com/mozilla/activity-stream/blob/8e75fd7690d9625d17f2b217122bce0f7076e998/lib/SimpleStorage.js#L28)):*
> ```JavaScript
> 
> const SIMPLE_STORAGE_QUOTA = "as-simple-storage.quota";
> // const QUOTA_DEFAULT = 20971520; // 20 MiB
> ```

please remove comment

---

*[lib/SimpleStorage.js, line 30 \[r1\]](https://reviewable.io:443/reviews/mozilla/activity-stream/1064#-KOpXwFxNZ_tYCynx0-L:-KOpXwFxNZ_tYCynx0-M:b-kuopd5) ([raw file](https://github.com/mozilla/activity-stream/blob/8e75fd7690d9625d17f2b217122bce0f7076e998/lib/SimpleStorage.js#L30)):*
> ```JavaScript
> // const QUOTA_DEFAULT = 20971520; // 20 MiB
> 
> const JETPACK_DIR_BASENAME = "jetpack";
> ```

we should use another base dirname to avoid collision. as-is, this will clobber simple-storage's JSON file for the addon

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mozilla/activity-stream/1064)*

**Comment by oyiptong** [comment]
Reviewed 4 of 4 files at r2.
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/mozilla/activity-stream/1064)*

**Comment by ncloudioj** [comment]
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[lib/MetadataCache.js, line 46 \[r2\]](https://reviewable.io:443/reviews/mozilla/activity-stream/1064#-KOpekX6Qvzg65lDo6wj:-KOphHzdCxAOWScNe37D:bf4vbn8) ([raw file](https://github.com/mozilla/activity-stream/blob/cd4f23b153b0e4e89ff4212340f35ea27730dfd4/lib/MetadataCache.js#L46)):*
<details><summary><i>Previously, oyiptong (Olivier Yiptong) wrote…</i></summary>
> each insert of a key in the list will be O(2n), which makes _moveToFront fairly expensive.
> Basically, each `get` costs O(n) where n is the number of elements in the list, twice. The first is upon`splice`, the other is upon `unshift`.
> 
> I would recommend either using a different data structure for the ordering (perhaps using a heap or linked list), or better, batching up the eviction.
> You could record the last access time for each key, and when the need to evict arises, throw away the items that have been the least recently used.
</details>
Yes, using JS array is not preferred here. The linked list would be ideal. Any existing library we could use? Because I also don't want to bloat it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mozilla/activity-stream/1064)*

**Comment by ncloudioj** [comment]
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[lib/SimpleStorage.js, line 30 \[r1\]](https://reviewable.io:443/reviews/mozilla/activity-stream/1064#-KOpXwFxNZ_tYCynx0-L:-KOpnLCNl0Vn1sW7cGrf:b7d9tw) ([raw file](https://github.com/mozilla/activity-stream/blob/8e75fd7690d9625d17f2b217122bce0f7076e998/lib/SimpleStorage.js#L30)):*
<details><summary><i>Previously, oyiptong (Olivier Yiptong) wrote…</i></summary>
> we should use another base dirname to avoid collision. as-is, this will clobber simple-storage's JSON file for the addon
</details>
Yes, it actually uses a different sub-dir for this json file. It should not leave the standard one alone.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mozilla/activity-stream/1064)*

File: .eslintignore [hunk 1]
```diff
@@ -3,3 +3,5 @@ logs/
 firefox/
 lib/vendor.bundle.js
 activity-streams-env/
+lib/SimpleStorage.js
+test/test-SimpleStorage.js
```
Comment (pdehaan) [inline .eslintignore]: Why excluding the tests?

File: test/test-MetadataCache.js [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+const fixtures = [
+  {desc: "It shuold hit", key: "foo", getKey: "foo", value: 123},
+  {desc: "It should miss", key: "bar", getKey: "nope", value: null},
+  {desc: "It should overwrite", key: "foo", getKey: "foo", value: 100}
+];
+
+const gMetadataCache = new MetadataCache();
+
+exports.test_cache_get = function*(assert) {
```
Comment (pdehaan) [inline test/test-MetadataCache.js]:
```
/home/travis/build/mozilla/activity-stream/test/test-MetadataCache.js
  16:26  error  This generator function does not have 'yield'  require-yield
  25:29  error  This generator function does not have 'yield'  require-yield
  36:35  error  This generator function does not have 'yield'  require-yield
  47:28  error  This generator function does not have 'yield'  require-yield
✖ 4 problems (4 errors, 0 warnings)
```

File: test/test-MetadataCache.js [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+  for (let fixture of fixtures) {
+    gMetadataCache.add(fixture.key, fixture.value);
+    let val = gMetadataCache.get(fixture.getKey);
+    assert.equal(fixture.value, val, fixture.desc);
+  }
+  assert.equal(gMetadataCache.length, 2);
+};
+
+exports.test_cache_remove = function*(assert) {
```
Comment (pdehaan) [inline test/test-MetadataCache.js]:
```
/home/travis/build/mozilla/activity-stream/test/test-MetadataCache.js
  25:29  error  This generator function does not have 'yield'  require-yield
```

File: test/test-MetadataCache.js [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+    gMetadataCache.add(fixture.key, fixture.value);
+  }
+  gMetadataCache.remove("nonsence");
+  assert.equal(gMetadataCache.length, 2, "It should remove anything if the to-be-removed key is missing");
+
+  gMetadataCache.remove("foo");
+  assert.ok(!gMetadataCache.get("foo"), "It should have removed the given key");
+};
+
+exports.test_cache_removeoldest = function*(assert) {
```
Comment (pdehaan) [inline test/test-MetadataCache.js]:
```
/home/travis/build/mozilla/activity-stream/test/test-MetadataCache.js
  36:35  error  This generator function does not have 'yield'  require-yield
```

File: test/test-MetadataCache.js [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+  assert.ok(!gMetadataCache.length, 0, "It should not remove anything if the cache is empty");
+
+  for (let fixture of fixtures) {
+    gMetadataCache.add(fixture.key, fixture.value);
+  }
+  gMetadataCache.removeOldest();
+  assert.ok(!gMetadataCache.get("bar"), "It should removed this key as the oldest one");
+};
+
+exports.test_pref_change = function*(assert) {
```
Comment (pdehaan) [inline test/test-MetadataCache.js]:
```
/home/travis/build/mozilla/activity-stream/test/test-MetadataCache.js
  47:28  error  This generator function does not have 'yield'  require-yield
```

File: .eslintignore (hunk 1 above)
Comment (ncloudioj) [inline .eslintignore]: We pulled these two files directly out from the addon sdk, looks like eslint was not happy with the code style there. Like one-liner if statements, and it threw out a lot of lint errors.

File: test/test-MetadataCache.js (hunk 3 above)
Comment (ncloudioj) [inline test/test-MetadataCache.js]: Yes, i have noticed these errors on Travis. Thanks for pointing out.

File: test/test-MetadataCache.js (hunk 2 above)
Comment (ncloudioj) [inline test/test-MetadataCache.js]: Good idea. Will do.

Comment (Mardak) [inline test/test-MetadataCache.js]: lint: `exports.test_cache_get = assert => {`

File: test/test-MetadataCache.js (hunk 3 above)
Comment (Mardak) [inline test/test-MetadataCache.js]: lint: `exports.test_cache_remove = assert => {`

File: test/test-MetadataCache.js [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+    // if (cache.length == 1 || ticked++ > 5) {
+      // return true;
+    // }
+    // return false;
+  // }, 1000);
+  // assert.equal(cache.length, 1, "It should have evicted items if it reaches the quota");
+  // simplePrefs.prefs["as-simple-storage.quota"] = oldQuota;
+// };
+
+after(exports, function() {
```
Comment (Mardak) [inline test/test-MetadataCache.js]: lint: `after(exports, () => {`

**Comment by oyiptong** [comment]
Review status: 3 of 8 files reviewed at latest revision, 12 unresolved discussions, some commit checks failed.

---

*[lib/SimpleStorage.js, line 30 \[r1\]](https://reviewable.io:443/reviews/mozilla/activity-stream/1064#-KOpXwFxNZ_tYCynx0-L:-KOqCvOjWovFDBQ8CkyT:b-60jhsj) ([raw file](https://github.com/mozilla/activity-stream/blob/8e75fd7690d9625d17f2b217122bce0f7076e998/lib/SimpleStorage.js#L30)):*
<details><summary><i>Previously, ncloudioj (Nan Jiang) wrote…</i></summary>
> Yes, it actually uses a different sub-dir for this json file. It should not leave the standard one alone.
</details>
Ahh, got it, you append `as-simple-storage` to the dirname.

---

*[test/test-MetadataCache.js, line 16 \[r2\]](https://reviewable.io:443/reviews/mozilla/activity-stream/1064#-KOpX37FvVa3n-Ez77yV-r2-16:-KOqDtsfpgLE3eEeJC1Z:b-hrrivx) ([raw file](https://github.com/mozilla/activity-stream/blob/cd4f23b153b0e4e89ff4212340f35ea27730dfd4/test/test-MetadataCache.js#L16)):*
<details><summary><i>Previously, Mardak (Ed Lee) wrote…</i></summary>
> lint: `exports.test_cache_get = assert => {`
</details>
Yes for the fat arrow!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mozilla/activity-stream/1064)*

**Comment by oyiptong** [comment]
Reviewed 3 of 3 files at r3.
Review status: 5 of 8 files reviewed at latest revision, 13 unresolved discussions, some commit checks failed.

---

*[lib/MetadataCache.js, line 61 \[r4\]](https://reviewable.io:443/reviews/mozilla/activity-stream/1064#-KOqEPKuqYuZdbxdarnn:-KOqEPKvRWNjCMTh3oOp:b-qychdd) ([raw file](https://github.com/mozilla/activity-stream/blob/c387b730f89d61d327f5e449cd34955acf8e5530/lib/MetadataCache.js#L61)):*
> ```JavaScript
> 
>   _moveToFront(key) {
>     const node = this._listGetNode(key);
> ```

A `Map` would be of great use here.

A call to `_listGetNode` is O(n) where n is the number of nodes in the linked list.

Using a `Map`, you'd have an amortized O(1) access to a list node, given a key.

Given that it's a doubly-linked list, the deletion and insertion should both be O(1)

---

*[lib/MetadataCache.js, line 93 \[r4\]](https://reviewable.io:443/reviews/mozilla/activity-stream/1064#-KOqEs5RchYP_4Qlizyo:-KOqEs5RchYP_4Qlizyp:by68lqv) ([raw file](https://github.com/mozilla/activity-stream/blob/c387b730f89d61d327f5e449cd34955acf8e5530/lib/MetadataCache.js#L93)):*
> ```JavaScript
> 
>   _removeElementByKey(key) {
>     const node = this._listGetNode(key);
> ```

As above, the use of a `Map` would be judicious here.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mozilla/activity-stream/1064)*

**Comment by ncloudioj** [comment]
@oyiptong Oh, you mean here we use another auxiliary map to search the node in the linked list. Sounds like a plan. let's push it to the extreme at the cost of some memory :)

File: test/test-MetadataCache.js (hunk 6 above)
Comment (ncloudioj) [inline test/test-MetadataCache.js]: Copy that! Thanks @Mardak! You use some robots to capture those lint errors, right 😄

**Comment by oyiptong** [comment]
Reviewed 3 of 3 files at r4.
Review status: 3 of 8 files reviewed at latest revision, 10 unresolved discussions.

---

*[package.json, line 10 \[r6\]](https://reviewable.io:443/reviews/mozilla/activity-stream/1064#-KOu6pswxSBKoaEhe-4s:-KOu6psxutWi2-1ACJKK:b-7x2hms) ([raw file](https://github.com/mozilla/activity-stream/blob/673c83d71901e85e58d35321ba2ee2340440a1c5/package.json#L10)):*
> ```JSON
>   },
>   "dependencies": {
>     "DoublyLinkedList": "github:andrewrjones/doubly-linked-list-js",
> ```

Something to consider: since it's not on NPM, this package could disappear anytime.

Consider either vendoring the module or using a package available on NPM. Leaving this as-is is good as well, since the data structure is easy to implement, but just something to keep in mind.

---

*[lib/MetadataCache.js, line 44 \[r6\]](https://reviewable.io:443/reviews/mozilla/activity-stream/1064#-KOuCEKmcGWCOp_lWiKk:-KOuCEKnQXeI6yRH98LA:bgatlxl) ([raw file](https://github.com/mozilla/activity-stream/blob/673c83d71901e85e58d35321ba2ee2340440a1c5/lib/MetadataCache.js#L44)):*
> ```JavaScript
>     if (key in this._cache) {
>       this._moveToFront(key);
>       this._cache[key] = value;
> ```

you can remove the second early return statement and making the code marginally more legible:

```javascript
add(key, value) {
  ...
  if (key in this._cache) {
    this._moveToFront(key);
  } else {
    // or call a new function _addNodeByKey
    let node = this._list.prepend(key);
    this._keyToNode[key] = node;
  }
  this_cache[key] = value;
}
```

---

*[lib/MetadataCache.js, line 63 \[r6\]](https://reviewable.io:443/reviews/mozilla/activity-stream/1064#-KOuDsllk4xUamdffNyP:-KOuDsllk4xUamdffNyQ:b-v7a3ru) ([raw file](https://github.com/mozilla/activity-stream/blob/673c83d71901e85e58d35321ba2ee2340440a1c5/lib/MetadataCache.js#L63)):*
> ```JavaScript
>     if (node) {
>       node.remove();
>       let newNode = this._list.prepend(key);
> ```

this occurs elsewhere... i suppose you could have a function called `_addNodeByKey`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mozilla/activity-stream/1064)*

**Comment by oyiptong** [comment]
Reviewed 3 of 3 files at r5, 2 of 2 files at r6.
Review status: all files reviewed at latest revision, 10 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/mozilla/activity-stream/1064)*

**Comment by ncloudioj** [comment]
Review status: all files reviewed at latest revision, 10 unresolved discussions.

---

*[package.json, line 10 \[r6\]](https://reviewable.io:443/reviews/mozilla/activity-stream/1064#-KOu6pswxSBKoaEhe-4s:-KOuFspC0z9yBw4V6KE7:bm6lga3) ([raw file](https://github.com/mozilla/activity-stream/blob/673c83d71901e85e58d35321ba2ee2340440a1c5/package.json#L10)):*
<details><summary><i>Previously, oyiptong (Olivier Yiptong) wrote…</i></summary>
> Something to consider: since it's not on NPM, this package could disappear anytime.
> 
> Consider either vendoring the module or using a package available on NPM. Leaving this as-is is good as well, since the data structure is easy to implement, but just something to keep in mind.
> 
</details>
I did a quick audit on a couple of implementations on npm/github, i think this one makes a lot sense to me, also licensed by MIT. Let's just cross fingers for now :) 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mozilla/activity-stream/1064)*

**Comment by ncloudioj** [comment]
To sum up all the changes in this PR:

* fork the simple-storage from addon sdk to increase the quota without affecting other addons. Both memoizer and metadata store share a 20MB backing file (configured by pref)
* add a LRU cache to cache metadata store query results, it'll evict the least recent used items once the backing file hits the quota
* enable the query cache for the preview provider
* (Not yet implemented) the LRU cache maintains some stats internally, it could be useful if we make the stats available in the debug page

@oyiptong One more run plz?
@sarracini Could you take a quick look at the change in PreviewProvider.js plz?

File: lib/PreviewProvider.js [hunk 7]
```diff
@@ -209,11 +210,29 @@ PreviewProvider.prototype = {
...
   /**
-   * Find the metadata for each link in the database
+   * Find the metadata for each link in the database. Note that it'll try to fetch items
+   * from the cache before querying the metadata store
    */
   _asyncFindItemsInDB: Task.async(function*(links) {
-    const cacheKeyArray = links.map(link => link.cache_key);
-    let linksMetadata = yield this._metadataStore.asyncGetMetadataByCacheKey(cacheKeyArray);
+    let cacheMissedKeys = [];
+    let linksMetadata = [];
```
Comment (sarracini) [inline lib/PreviewProvider.js]: maybe renamed ```linksMetadata``` to ```existingLinksMetadata``` and then just make another ```linksMetadata``` which will be the one you return

File: lib/PreviewProvider.js [hunk 8]
```diff
@@ -209,11 +210,29 @@ PreviewProvider.prototype = {
...
+      if (fromCache) {
+        linksMetadata.push(fromCache);
+      } else {
+        cacheMissedKeys.push(key);
+      }
+    });
+
+    // Hit the database for the missing keys
+    if (cacheMissedKeys.length) {
+      const fromStore = yield this._metadataStore.asyncGetMetadataByCacheKey(cacheMissedKeys);
```
Comment (sarracini) [inline lib/PreviewProvider.js]: maybe renamed ```fromStore``` to ```missedLinksMetadata``` to be more clear

File: lib/PreviewProvider.js (hunk 7 above)
Comment (sarracini) [inline lib/PreviewProvider.js]: actually nvm leave it

File: lib/PreviewProvider.js [hunk 9]
```diff
@@ -209,11 +210,29 @@ PreviewProvider.prototype = {
...
+      const fromCache = MetadataCache.cache.get(key);
+      if (fromCache) {
+        linksMetadata.push(fromCache);
+      } else {
+        cacheMissedKeys.push(key);
+      }
+    });
+
+    // Hit the database for the missing keys
+    if (cacheMissedKeys.length) {
```
Comment (sarracini) [inline lib/PreviewProvider.js]: can you be explicit here and do ```cachedMissedKeys.length > 0``` (you never know with js)

Comment (ncloudioj) [inline lib/PreviewProvider.js]: heh, good idea. Be safe.

File: lib/PreviewProvider.js [hunk 10]
```diff
@@ -209,11 +210,29 @@ PreviewProvider.prototype = {
...
+    let linksMetadata = [];
+
+    // Hit the cache first
+    links.forEach(link => {
+      const key = link.cache_key;
+      const fromCache = MetadataCache.cache.get(key);
+      if (fromCache) {
+        linksMetadata.push(fromCache);
+      } else {
+        cacheMissedKeys.push(key);
```
Comment (sarracini) [inline lib/PreviewProvider.js]: just to be sure, ```MetadataCache.cache.get(key)``` will return just the link's cache_key right? In other words you don't need to make a ```cacheKeyArray``` right?

Comment (sarracini) [inline lib/PreviewProvider.js]: oh nevermind! read this wrong, all good :)

Comment (ncloudioj) [inline lib/PreviewProvider.js]: Yes, it returns the whole link just as what you've insert.

**Comment by sarracini** [comment]
@ncloudioj LGTM w/ nits on PreviewProvider side

**Comment by ncloudioj** [comment]
Review status: 4 of 10 files reviewed at latest revision, 13 unresolved discussions.

---

*[.eslintignore, line 7 \[r2\]](https://reviewable.io:443/reviews/mozilla/activity-stream/1064#-KOpX37FvVa3n-Ez77yT-r2-7:-KOzlpENuPR35wsXj4Wx:b-896fix) ([raw file](https://github.com/mozilla/activity-stream/blob/cd4f23b153b0e4e89ff4212340f35ea27730dfd4/.eslintignore#L7)):*
<details><summary><i>Previously, ncloudioj (Nan Jiang) wrote…</i></summary>
> We pulled these two files directly out from the addon sdk, looks like eslint was not happy with the code style there. Like one-liner if statements, and it threw out a lot of lint errors.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mozilla/activity-stream/1064)*

**Comment by oyiptong** [comment]
Review status: 4 of 10 files reviewed at latest revision, 11 unresolved discussions, some commit checks failed.

---

*[lib/MetadataCache.js, line 25 \[r9\]](https://reviewable.io:443/reviews/mozilla/activity-stream/1064#-KP-F-abBRzhHb1gZGgp:-KP-F-abBRzhHb1gZGgq:b4jhtuv) ([raw file](https://github.com/mozilla/activity-stream/blob/d06b17b747ea28ac68cf62e0343d8dc9ca1862bc/lib/MetadataCache.js#L25)):*
> ```JavaScript
>   });
>   this._onEvicted = onEvicted;
>   this._hits = 0;
> ```

I would leave out the hit/miss code for now and implement this as a larger feature later

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mozilla/activity-stream/1064)*

**Comment by oyiptong** [comment]
Reviewed 2 of 2 files at r7, 2 of 4 files at r8, 2 of 2 files at r9.
Review status: all files reviewed at latest revision, 11 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/mozilla/activity-stream/1064)*

**Comment by oyiptong** [comment]
:lgtm: w/ removing the cache hit/miss accounting

---

Review status: all files reviewed at latest revision, 11 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/mozilla/activity-stream/1064#-:-KP-FDUYEGHDhrfNQ5tG:bkk4qyy)*

**Comment by oyiptong** [comment]
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[lib/PreviewProvider.js, line 232 \[r9\]](https://reviewable.io:443/reviews/mozilla/activity-stream/1064#-KOv_PmLlFRprvvzNsGG-r9-232:-KP-GCK9JHVr3Ag0pXsw:b-es01hv) ([raw file](https://github.com/mozilla/activity-stream/blob/d06b17b747ea28ac68cf62e0343d8dc9ca1862bc/lib/PreviewProvider.js#L232)):*
<details><summary><i>Previously, ncloudioj (Nan Jiang) wrote…</i></summary>
> heh, good idea. Be safe.
</details>
I think in this case, expecting a truthy value for `cachedMissedKey.length` is ok

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mozilla/activity-stream/1064)*

**Pull request merged**